### DISCUSSION TITLE
HowTo detect errors in database layer?

### DISCUSSION THREAD

**Discussion by hoeverma** [serverpod/serverpod#2160] [Q&A] [answered]
The database layer functions in Serverpod (insert, update, ...) always return (list of) objects. How can errors be detected (e.g.: call update for a not existing object, or anything goes wrong in Postgres internally)? Do these functions throw exceptions?

**Comment by bfemmer** [comment, marked as answer]
You should probably design/define your own meaningful exceptions and then throw them from the server (catching them in your app so that you can handle user state and/or user messaging appropriate for the circumstance). The documentation (found here: https://docs.serverpod.dev/concepts/exceptions) will send you down the right path.

**Reply by hoeverma** [reply]
Thank you very much for taking the time to answer! Maybe I am a bit dumb there, but just to be completely sure: The CRUD database functions on the server (db.update, db.insertRow, ...) indicate an error by throwing an exception. If no exception is thrown, the operation succeeded. Is this right?

**Reply by SandPod** [reply]
That is correct! 👍 No exception is proof of a successful operation.

**Comment by hoeverma** [comment]
> That is correct! 👍 No exception is proof of a successful operation.

**Comment by hoeverma** [comment]
Thank you!
